### ISSUE TITLE
.dockerignore ignores trailing slash on directories

### ISSUE THREAD

**Issue by mikexstudios** [docker/docker-py#581]
Since docker 1.4.0, `.dockerignore` files became `.gitignore`-like in that there is a difference between `foo` vs. `foo/`. The former can match files, symbolic links, etc. while the latter only matches the directory `foo`. This problem was also reported in: https://github.com/docker/compose/issues/543.

Unfortunately, docker-py cannot handle `foo/` syntax, and I suspect it's missing some other major cases. The reason is here: https://github.com/docker/docker-py/blob/0b78d7a841f123b61a1e03cec7dfc29c92dde0fe/docker/utils/utils.py#L67-L68. docker-py uses directories from `os.walk` which output directories without trailing slash and `fnmatch` can't match:
```python
>>> from fnmatch import fnmatch
>>> fnmatch('src', 'src/')
False
```

My first thought was to suggest that docker-py use pathspec (https://github.com/cpburnz/python-path-specification) which implements .gitignore-style matching. However, I don't think the way that docker implements .dockerignore matching mirrors .gitignore-style matching (see: https://github.com/docker/docker/blob/master/pkg/fileutils/fileutils.go).

Does anyone have good ideas on an elegant fix?

As a temporary fix to anyone who stumbles on this issue, try to keep your `.dockerignore` file as simple as possible if you find that certain directories are not being ignored.

**Comment by shin-** [comment]
Hey @mikexstudios , thank you for reporting this!

I should have a patch for this by the end of the day, do you think you could help me test it with a real-life `.dockerignore` file?

**Comment by mikexstudios** [comment]
Hey @shin-. I updated your PR with a unit test for this case. Your PR works for detecting trailing slashes on folders.

However, I am concerned that we need a more general and robust way of handling .dockerignore files. The currently implementation fails on a lot of syntax defined here: https://github.com/docker/docker/blob/master/docs/sources/reference/builder.md#dockerignore-file

I'll see if I can contribute to improving this in docker-py.

**Comment by aanand** [comment]
Closed by #721.
